transaction/core/src/membership_proofs/mod.rs

```diff
@@ -66,96 +201,38 @@ pub fn is_membership_proof_valid(
+        .first()
+        .ok_or_else(|| Error::MissingLeafHash(proof.index))?;
+    if first.range.from != proof.index || first.range.to != proof.index {
+        return Err(Error::MissingLeafHash(proof.index));
     }
-
-    // * The root hash of the proof must match the known root_hash.
-    let mut ranges: Vec<&Range> = range_to_hash.keys().collect();
-    ranges.sort();
-
-    let root_range = ranges.last().expect("`ranges` should be non-empty.");
-    let root_hash = range_to_hash
-        .get(root_range)
-        .expect("`root_range` should be a key.");
-
-    if *root_hash != *known_root_hash {
-        // Incorrect root hash.
-        return Ok(false);
+    // The first element hash must match the tx_out
+    if !bool::from(first.hash.as_ref().ct_eq(&hash_leaf(tx_out))) {
+        return Err(Error::IncorrectLeafHash(first.range.from));
     }
 
-    if proof.highest_index > root_range.to {
-        return Ok(false);
-    }
+    // Compute the implied root hash, or an error if this can't be done
+    // This should have fixed access patterns regardless of input, except on erroring code paths.
+    let implied_root = compute_implied_merkle_root(proof)?;
 
-    // * All internal node's hashes between leaf and root must be recomputable from their children's hashes.
-    let ranges_containing_tx_out: Vec<&Range> = ranges
-        .iter()
-        .cloned()
-        .filter(|range| range.from <= proof.index && proof.index <= range.to)
-        .collect();
-
-    for range in &ranges_containing_tx_out {
-        let hash = range_to_hash
-            .get(range)
-            .expect("range_to_hash must contain range");
-        if range.from != range.to {
-            // Internal Node.
-            let mid: u64 = (range.from + range.to) / 2;
-
-            // Left child.
-            let left_child_range = Range::new(range.from, mid)?;
-            let left_child_hash = match range_to_hash.get(&left_child_range) {
-                Some(hash) => hash,
-                None => {
-                    // Proof does not contain a required hash.
-                    return Ok(false);
-                }
-            };
-
-            // Right child.
-            let right_child_range = Range::new(mid + 1, range.to)?;
-            let right_child_hash = match range_to_hash.get(&right_child_range) {
-                Some(hash) => hash,
-                None => {
-                    // Proof does not contain a required hash.
-                    return Ok(false);
-                }
-            };
-
-            let expected_hash = hash_nodes(left_child_hash, right_child_hash);
-            if *hash != expected_hash {
-                // Proof contains an incorrect hash value.
-                return Ok(false);
-            }
-        }
+    // This test is carried over from earlier revisions
+    // TODO: Should this be a test for equality?
+    // TODO: Should we also test that implied_root.range.from == 0?
```

Comment:
What's needed to resolve these TODOs?

Reply:
(this line is where that test came back)

Reply:
to resolve these TODO's, I need more context on all the places this is used I guess -- is there some way that a valid merkle proof is generated that doesn't have implied_root.range.from == 0 ?
can using the derived merkle proof thing cause that?
we can just try it and see if it passes tests
we can ask matt if he sees a problem with that or a reason not to do it?
we can try to read more of the code again to figure this out

probably all of the above :)

Reply:
ok, i think the highest_index is not supposed to be the same as merkle_root.to, it's supposed to be equal to index i think

i think this is resolved now